README.md
# `ilustrado`

## Summary

`ilustrado` is a Python package that implements a highly-customisable massively-parallel genetic algorithm for *ab initio* crystal structure prediction (CSP), with a focus on mapping out compositional phase diagrams. The aim of `ilustrado` was to provide a method of extending CSP results generated with random structure searching (AIRSS) or extrapolating from known chemically-relevant systems (species-swapping/prototyping).

The API is [fully-documented](http://www.tcm.phy.cam.ac.uk/~me388/ilustrado) and the source code can be found on [BitBucket](https://bitbucket.org/me388/ilustrado). `ilustrado` makes extensive use of the [matador](https://tcm.phy.cam.ac.uk/~me388/matador) API and interfaces with [CASTEP](http://www.castep.org/) for DFT-level relaxations. Written by [Matthew Evans](http://www.tcm.phy.cam.ac.uk/~me388) (me388@cam.ac.uk).

By default, fitnesses are evaluated as the distance from a binary or ternary convex hull that is passed as input. Duplicate structures are filtered out post-relaxation based on pair distribution function overlap. The standard mutation operators are implemented: cell and position noise, vacancies, and atom permutation and transmutation. Additionally a Voronoi-based mutation has been implemented:
1. Each of the *N* atoms of randomly chosen species **A** are removed from the cell.
2. The Voronoi decomposition of the remaining atoms is calculated.
3. K-means clustering is applied to the Voronoi points to split them into *N +/- D* clusters, where *D* is a random integer less than *sqrt(N)*.
4. *N +/- D* atoms of species **A** are added to the cell at the centres of these clusters.
The species **A** can be specified by the user or chosen randomly. This mutation is effective when studying materials that have one light, mobile element, for example Li. 

Crossover is performed with the standard cut-and-splice method [1] to ensure transferrability over many types of material systems. This method cuts a random fraction from each parent cell and splices them back together; in `ilustrado` efforts are made to align the cells such that the cutting axes are approximately commensurate before crossover.

Several physical constraints (minimum atomic separations, densities, cell shapes) are applied to the trial structures before relaxation to improve efficiency. In order to maintain population diversity as the simulation progresses, the user can optionally disfavour frequently-visited regions of composition space.

The entrypoint is a Python script that creates an `ArtificialSelector` object with the desired parameters (documented [here](http://www.tcm.phy.cam.ac.uk/~me388/ilustrado/ilustrado.html#ilustrado.ilustrado.ArtificialSelector)). Many examples can be found in `examples/`. There are two `compute_mode`s in which to run `ilustrado`.
- `compute_mode='direct'` involves one `ilustrado` processes spawning `mpirun` calls either on local or remote partitions (i.e. either a node list is passed for running on a local cluster, or `ilustrado` itself is submitted as a HPC job). In this case, the user must manually restart the GA when the job finishes.
- `compute_mode='slurm'` performs the GA in interruptible steps; submitting `ilustrado` as a job will lead to the submission of many slurm array jobs for the relaxation, and a dependency job that re-runs the `ilustrado` process to check on the relaxations. In this case, the user only needs to submit one job (tested on 6400 cores/200 nodes without issue).

[1] Deaven, D. M.; Ho, K. M. Molecular Geometry Optimization with a Genetic Algorithm. Phys. Rev. Lett. 1995, 75, 288, [10.1103/PhysRevLett.75.288](https://doi.org/10.1103/PhysRevLett.75.288)


## New in v0.3b:

- `sandbagging` of composition space
- `compute_mode='slurm'` that makes use of array jobs for "infinite" horizontal scalability
- improved documentation and examples

## API Docs

- [Auto-generated API docs](http://www.tcm.phy.cam.ac.uk/~me388/ilustrado/ilustrado.html)


CHANGELOG.rst
v0.4
====

- Updated matador dependency PyPI versions > 0.9.
- Added ``compute_mode="manual"`` to just generate structures without relaxing
- Added GitHub CI
- Rescale cell volume after crossover to maintain density
- QoL updates to printing and loading mutations
- Allow arbitrary ASE calculator to be used in relaxations

v0.3
====

- New keyword: ``sandbagging``. When enabled, fitness penalties will be applied to successive sampling of the same region of composition space. By default, the modifier is a multiplicative factor of 0.95 to all compositions within a hypersphere of radius 0.05.
- ``compute_mode='slurm'`` that makes use of array jobs for "infinite" horizontal scalability
- improved documentation and examples

.. |MIT License| image:: https://img.shields.io/badge/license-MIT-blue.svg
   :target: https://github.com/ml-evs/ilustrado/blob/master/LICENSE
.. |Documentation Status| image:: https://readthedocs.org/projects/ilustrado/badge/?version=latest
   :target: https://ilustrado.readthedocs.io/en/latest/?badge=latest


ilustrado/fitness.py
# coding: utf-8
""" This file implements all notions of fitness. """
import numpy as np
from matador.utils.cursor_utils import get_array_from_cursor
from matador.utils.chem_utils import get_concentration, get_formation_energy


class FitnessCalculator:
    """ This class calculates the fitnesses of generations,
    by some global definition of generation-agnostic fitness.

    Parameters:
        fitness_metric (str): either 'dummy', 'hull' or 'hull_test'.
        fitness_function (callable): function to operate on numpy array of raw fitness values,
        hull (QueryConvexHull): matador hull from which to calculate metastability,
        sandbagging (bool): whether or not to "sandbag" particular compositions, i.e. lower
            a structure's fitness based on the number of nearby phases

    """

    def __init__(
        self,
        fitness_metric="dummy",
        fitness_function=None,
        hull=None,
        sandbagging=False,
        debug=False,
    ):
        """ Initialise fitness calculator, if from hull then
        extract chemical potentials.

        """
        self.testing = False
        self.debug = debug
        self.fitness_metric = fitness_metric
        if self.fitness_metric == "hull":
            self._get_raw = self._get_hull_distance
            if hull is None:
                raise RuntimeError("Cannot calculate hull distance without a hull!")
            self.hull = hull
            self.chempots = hull.chempot_cursor
        elif self.fitness_metric == "dummy":
            self._get_raw = self._get_dummy_fitness
        elif self.fitness_metric == "hull_test":
            self.testing = True
            self._get_raw = self._get_hull_distance
            self.hull = hull
            self.chempots = hull.chempot_cursor
        else:
            raise RuntimeError("No recognised fitness metric given.")

        self.sandbagging = False
        if sandbagging:
            self.sandbagging = True
            self.sandbag_multipliers = dict()

        if fitness_function is None:
            self.fitness_function = default_fitness_function
        else:
            self.fitness_function = fitness_function

    def evaluate(self, generation):
        """ Assign normalised fitnesses to an entire generation.
        Normalisation uses the logistic function such that

        `fitness = 1 - tanh(2*distance_from_hull)`,

        Parameters:
            generation (Generation/list): list/iterator over optimised structures,

        """

        raw = self._get_raw(generation)
        fitnesses = self.fitness_function(raw)

        for ind, _ in enumerate(generation):
            generation[ind]["raw_fitness"] = raw[ind]
            generation[ind]["fitness"] = fitnesses[ind]

        if self.sandbagging:
            self.update_sandbag_multipliers(generation)
            self.apply_sandbag_multipliers(generation)

    def update_sandbag_multipliers(self, generation, modifier=0.95):
        """ Assign composition penalty based on number of nearby structures.
        Updates fitness.sandbag_multipliers to a dictionary with chemical concentration as keys
        and values of fitness penalty.

        Parameters:
            generation (Generation): list of optimised structures.

        """
        for structure in generation:
            if tuple(structure["concentration"]) in self.sandbag_multipliers:
                self.sandbag_multipliers[tuple(structure["concentration"])] *= modifier
            else:
                self.sandbag_multipliers[tuple(structure["concentration"])] = modifier

    def apply_sandbag_multipliers(self, generation, locality=0.05):
        """ Scale the generation's fitness by the sandbag modifier. This
        updates the 'fitness' key and the 'modifier' key (total scaling) of each document
        in the generation.

        Parameters:
            generation (Generation): list of optimised structures.

        Keyword Arguments:
            locality (float): tolerance by which two structures are "nearby"

        """
        for ind, structure in enumerate(generation):
            generation[ind]["modifier"] = 1
            for concentration in self.sandbag_multipliers:
                if (
                    np.sqrt(
                        np.sum(
                            np.abs(
                                np.asarray(structure["concentration"])
                                - np.asarray(list(concentration))
                            )
                            ** 2
                        )
                    )
                    <= locality
                ):
                    generation[ind]["modifier"] *= self.sandbag_multipliers[
                        concentration
                    ]
            generation[ind]["fitness"] *= generation[ind]["modifier"]

    def _get_hull_distance(self, generation):
        """ Assign distance from the hull from hull for generation,
        assigning it.

        Parameters:
            generation (Generation): list of optimised structures.

        Returns:
            hull_dist (list(float)): list of distances to the hull.

        """
        for ind, populum in enumerate(generation):
            generation[ind]["concentration"] = get_concentration(
                populum, self.hull.elements
            )
            generation[ind]["formation_enthalpy_per_atom"] = get_formation_energy(
                self.chempots, populum
            )
            if self.debug:
                print(
                    generation[ind]["concentration"],
                    generation[ind]["formation_enthalpy_per_atom"],
                )
        if self.testing:
            for ind, populum in enumerate(generation):
                generation[ind]["formation_enthalpy_per_atom"] = np.random.rand() - 0.5
        structures = np.hstack(
            (
                get_array_from_cursor(generation, "concentration"),
                get_array_from_cursor(
                    generation, "formation_enthalpy_per_atom"
                ).reshape(len(generation), 1),
            )
        )
        if self.debug:
            print(structures)
        hull_dist = self.hull.get_hull_distances(structures, precompute=False)
        for ind, populum in enumerate(generation):
            generation[ind]["hull_distance"] = hull_dist[ind]
        return hull_dist

    def _get_dummy_fitness(self, generation):
        """ Generate dummy hull distances from -0.01 to 0.05.

        Parameters:
            generation (Generation): list of optimised structures.

        Returns:
            list(float): dummy list of hull distances.

        """
        return (0.05 * np.random.rand(len(generation)) - 0.01).tolist()


def default_fitness_function(raw, c=50, offset=0.075):
    """ Default fitness function: logistic function.

    Parameters:
        raw (ndarray): 1D array of raw fitness values.

    Returns:
        ndarray: 1D array of rescaled fitnesses.

    """
    fitnesses = 1 / (1 + np.exp(c * (raw - offset)))
    if isinstance(fitnesses, np.float64):
        fitnesses = min(1, fitnesses)
    else:
        fitnesses[fitnesses > 1.0] = 1.0
    return fitnesses
